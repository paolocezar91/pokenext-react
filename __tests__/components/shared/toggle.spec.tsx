import Toggle from "@/components/shared/toggle";
import "@testing-library/jest-dom";
import { fireEvent, render, screen } from "@testing-library/react";

describe("Toggle Component", () => {
  const mockOnChange = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render with default unchecked state", () => {
    render(<Toggle id="test" value={false} onChange={mockOnChange} />);

    const checkbox = screen.getByRole("checkbox") as HTMLInputElement;
    expect(checkbox.checked).toBeFalsy();
    expect(checkbox).not.toBeChecked();
  });

  it("should render with checked state when value is true", () => {
    render(<Toggle id="test" value onChange={mockOnChange} />);

    const checkbox = screen.getByRole("checkbox");
    expect(checkbox).toBeTruthy();
    expect(checkbox).toBeChecked();
  });

  it("should display children text when provided", () => {
    render(
      <Toggle
        id="test"
        value={false}
        onChange={mockOnChange}
        childrenLeft="Test Label"
      ></Toggle>,
    );
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should display className when provided", () => {
    render(
      <Toggle
        id="test"
        className="test-class"
        value={false}
        onChange={mockOnChange}
        childrenLeft="Test Label"
      ></Toggle>,
    );
    expect(screen.getByTestId("toggle-label")).toHaveClass("test-class");
  });

  it("should call onChange with true when toggling from off to on", () => {
    render(<Toggle id="test" value={false} onChange={mockOnChange} />);

    const checkbox = screen.getByRole("checkbox");
    fireEvent.click(checkbox);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(true);
  });

  it("should call onChange with false when toggling from on to off", () => {
    render(<Toggle id="test" value onChange={mockOnChange} />);

    const checkbox = screen.getByRole("checkbox");
    fireEvent.click(checkbox);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(false);
  });

  it("should toggle when clicking on the label text", () => {
    render(
      <Toggle
        id="test"
        value={false}
        onChange={mockOnChange}
        childrenLeft="Toggle Me"
      ></Toggle>,
    );

    const labelText = screen.getByText("Toggle Me");
    fireEvent.click(labelText);

    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith(true);
  });

  it("should have proper accessibility attributes", () => {
    render(
      <Toggle
        id="test"
        value={false}
        onChange={mockOnChange}
        childrenLeft="Accessible Toggle"
      ></Toggle>,
    );

    const checkbox = screen.getByRole("checkbox");
    expect(checkbox).toHaveAttribute("type", "checkbox");
    expect(checkbox).toHaveClass("sr-only peer");
  });

  it("should match snapshot when unchecked", () => {
    const { asFragment } = render(
      <Toggle id="test" value={false} onChange={mockOnChange} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot when checked", () => {
    const { asFragment } = render(
      <Toggle id="test" value onChange={mockOnChange} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with children", () => {
    const { asFragment } = render(
      <Toggle
        id="test"
        value={false}
        onChange={mockOnChange}
        childrenLeft="Label"
      ></Toggle>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
