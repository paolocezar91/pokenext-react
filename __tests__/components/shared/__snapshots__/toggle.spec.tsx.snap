// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Component should match snapshot when checked 1`] = `
<DocumentFragment>
  <label
    class=" inline-flex items-center cursor-pointer"
    data-testid="toggle-label"
    for="toggle-test"
  >
    <input
      aria-label="Toggle"
      checked=""
      class="sr-only peer"
      id="toggle-test"
      type="checkbox"
    />
    <div
      class="
      relative
      bg-gray-700
      border-gray-600
      rounded-full
      rtl:peer-checked:after:-translate-x-full
      peer
      peer-focus:outline-none
      peer-focus:ring-2
      peer-focus:ring-(--pokedex-blue)
      peer-focus:ring-white
      peer-checked:after:translate-x-full
      peer-checked:after:border-(--pokedex-red-light)
      peer-checked:bg-(--pokedex-blue)
      after:transition-all
      after:content-['']
      after:absolute
      after:top-[2px]
      after:start-[2px]
      after:bg-white
      after:border-gray-300
      after:border
      after:rounded-full
      w-11 h-6 after:h-5 after:w-5
    "
    />
  </label>
</DocumentFragment>
`;

exports[`Toggle Component should match snapshot when unchecked 1`] = `
<DocumentFragment>
  <label
    class=" inline-flex items-center cursor-pointer"
    data-testid="toggle-label"
    for="toggle-test"
  >
    <input
      aria-label="Toggle"
      class="sr-only peer"
      id="toggle-test"
      type="checkbox"
    />
    <div
      class="
      relative
      bg-gray-700
      border-gray-600
      rounded-full
      rtl:peer-checked:after:-translate-x-full
      peer
      peer-focus:outline-none
      peer-focus:ring-2
      peer-focus:ring-(--pokedex-blue)
      peer-focus:ring-white
      peer-checked:after:translate-x-full
      peer-checked:after:border-(--pokedex-red-light)
      peer-checked:bg-(--pokedex-blue)
      after:transition-all
      after:content-['']
      after:absolute
      after:top-[2px]
      after:start-[2px]
      after:bg-white
      after:border-gray-300
      after:border
      after:rounded-full
      w-11 h-6 after:h-5 after:w-5
    "
    />
  </label>
</DocumentFragment>
`;

exports[`Toggle Component should match snapshot with children 1`] = `
<DocumentFragment>
  <label
    class=" inline-flex items-center cursor-pointer"
    data-testid="toggle-label"
    for="toggle-test"
  >
    <input
      aria-label="Label"
      class="sr-only peer"
      id="toggle-test"
      type="checkbox"
    />
    <span
      class="ms-3 mr-2 text-xs text-foreground"
    >
       Label 
    </span>
    <div
      class="
      relative
      bg-gray-700
      border-gray-600
      rounded-full
      rtl:peer-checked:after:-translate-x-full
      peer
      peer-focus:outline-none
      peer-focus:ring-2
      peer-focus:ring-(--pokedex-blue)
      peer-focus:ring-white
      peer-checked:after:translate-x-full
      peer-checked:after:border-(--pokedex-red-light)
      peer-checked:bg-(--pokedex-blue)
      after:transition-all
      after:content-['']
      after:absolute
      after:top-[2px]
      after:start-[2px]
      after:bg-white
      after:border-gray-300
      after:border
      after:rounded-full
      w-11 h-6 after:h-5 after:w-5
    "
    />
  </label>
</DocumentFragment>
`;
